binderhub/tests/test_utils.py

```diff
@@ -0,0 +1,56 @@
+from binderhub import utils
+
+
+def test_rendezvous_rank():
+    # check that a key doesn't move if its assigned bucket remains but the
+    # other buckets are removed
+    key = "crazy frog is a crazy key"
+    first_round = utils.rendezvous_rank(["b1", "b2", "b3"], key)
+    second_round = utils.rendezvous_rank([first_round[0], first_round[1]], key)
+
+    assert first_round[0] == second_round[0], key
+
+
+def test_rendezvous_independence():
+    # check that the relative ranking of 80 buckets doesn't depend on the
+    # presence of 20 extra buckets
+    key = "k1"
+    eighty_buckets = utils.rendezvous_rank(["b%i" % i for i in range(80)], key)
+    hundred_buckets = utils.rendezvous_rank(["b%i" % i for i in range(100)], key)
+
+    for i in range(80, 100):
+        hundred_buckets.remove("b%i" % i)
+
+    assert eighty_buckets == hundred_buckets
+
+
+def test_rendezvous_redistribution():
+    # check that approximately a third of keys move to the new bucket
+    # when one is added
```

Comment:
An idea for how to test that keys move when a new bucket becomes available and that the pattern of movement is right. I think this is how it should be but not sure. WDYT?

Reply:
Excellent!

Since we actually hash "nodes-key", this test doesn't need to be run multiple times, but if we would hash node separately from key and as consistent hashing was described to do as compared to our rendezvous hashing, then we could by fluke have two node hashes that are spaced luckily allow for the new node to to catch 1/3 of the stuff.

But hmmm, could you position nodes like pointers on a clock to initially have a fair share and then also after have a fair share?

Thats a clean test as well to have I think, to check that we have a 1/2 distribution initially and then get a 1/3 distribution after, combined with the previous test about perfectly stable we capture all kinds of logic. You are already doing this to some degree but it is captures mostly by the `abs(from_b1 - from_b2) < 10` statement.